rust/arrow/src/array/builder.rs

```diff
@@ -253,8 +253,18 @@ pub trait BufferBuilderTrait<T: ArrowPrimitiveType> {
 }
 
 impl<T: ArrowPrimitiveType> BufferBuilderTrait<T> for BufferBuilder<T> {
-    default fn new(capacity: usize) -> Self {
-        let buffer = MutableBuffer::new(capacity * mem::size_of::<T::Native>());
+    #[inline]
+    fn new(capacity: usize) -> Self {
+        let buffer = if T::DATA_TYPE == DataType::Boolean {
```

Comment:
Note that after templating, this will be either `if a == a` or `if a != a`, which I am hoping the compiler to optimize out.

Reply:
I think using an `if` for specialization rather than the compiler is totally legit

Reply:
Nice! I used a similar trick for the simd sum aggregation that the compiler was also able to optimize away (https://github.com/apache/arrow/pull/8370/files#diff-53a8522c4a482c9566b9032e1bd2c7990bab29640857cb40411beb7158189e75R625)

Reply:
@jorgecarleitao 

I see this in profiling for BufferBuilderTrait append:

![image](https://user-images.githubusercontent.com/163737/101012554-4c1dfc00-3563-11eb-9b00-f92eac06b7ae.png)

Which accounts for 9% of the instruction fetches according to callgrind when running the TPHC 12 benchmark.

So it seems the `if T::DATA_TYPE == DataType::Boolean` is not being optimized out and goes into the partial eq implementation (or the profiler is wrong), maybe because the implementation there is not easy to optimize / specialize for this case. I will check if another implementation (pattern match?) solves this.